# Imports

In [11]:
## Suppress warnings
import warnings
warnings.filterwarnings("ignore")
import pandas as pd
import numpy as np
import seaborn as sns
import pickle

## Notebook display settings
from IPython.display import display
pd.options.display.max_columns = None
pd.options.display.max_rows = None
pd.options.display.max_seq_items = None

# Own scripts
from functions import get_start_end_points, generate_histogram2d, gen_spirals, analyse_spiral
## Plot style
%matplotlib inline

# Daten

In [12]:
with open('data.pickle', 'rb') as handle:
    data = pickle.load(handle)
data.head()

id  run               timestamp  segment          start_time  \
0  163.0  0.0 2017-05-30 09:21:13.150     11.0 2017-05-30 09:21:05   
1  164.0  0.0 2017-05-30 09:21:13.200     11.0 2017-05-30 09:21:05   
2  165.0  0.0 2017-05-30 09:21:13.250     11.0 2017-05-30 09:21:05   
3  166.0  0.0 2017-05-30 09:21:13.300     11.0 2017-05-30 09:21:05   
4  167.0  0.0 2017-05-30 09:21:13.350     11.0 2017-05-30 09:21:05   

                 end_time   force_x   force_y   force_z       pos_x  \
0 2017-05-30 09:22:41.607 -0.543045 -2.924596 -7.257000 -257.695704   
1 2017-05-30 09:22:41.607 -0.465969 -3.278856 -7.522000 -257.617629   
2 2017-05-30 09:22:41.607 -0.317475 -3.435834 -7.705999 -256.626803   
3 2017-05-30 09:22:41.607 -0.298383 -3.457754 -7.580999 -254.594017   
4 2017-05-30 09:22:41.607 -0.289898 -3.447854 -7.539000 -251.411919   

        pos_y       pos_z       rot_x       rot_y       rot_z result_robot  
0 -265.980105  240.007357   -0.008832    0.034723  179.958414      Success  
1 -265.999590  240.004745   -0.000171    0.019928  179.968161      Success  
2 -266.602650  239.791224 -179.993978  179.989304    0.164766      Success  
3 -267.926090  239.323013 -179.992478  179.999036    0.577883      Success  
4 -269.894497  238.550625 -179.996649  179.988091    1.210269      Success

# Code

## Listen

Start und Endpunkte der Such-Segmente finden

In [13]:
start_points, end_points, rel_end_points = get_start_end_points(data)

In [14]:
hist, xedge, yedge = generate_histogram2d(rel_end_points)

# Spiralen

In [15]:
spirals_r5 = gen_spirals(hist, xedge, yedge, -0.019, -0.107)
spirals_r10 = gen_spirals(hist, xedge, yedge, 0.007, -0.076)
spirals_r15 = gen_spirals(hist, xedge, yedge, 0.015, -0.065)
spirals_r20 = gen_spirals(hist, xedge, yedge, 0.019, -0.029)

## Definitionen

Zur besseren Verständlichkeit soll folgendes gelten:

- Norden (N) und Osten (E) sind die jeweiligen positiven Richtungen der x oder y Achse
- Süden (S) und Westen (W) sind die jeweiligen negativen Richtungen der x oder y Achse
- Die x Achse wird horizontal dargestellt, die Werte werden von niedrig/negativ (links) nach hoch/positiv (rechts) aufgetragen
- Die y Achse wird vertikal dargestellt, die Werte werden von niedrig/negativ (unten) nach hoch/positiv (oben) aufgetragen

## Analysis

In [16]:
_ = analyse_spiral(spirals_r5, hist)

direction.S turn.turn_left: 90180.0

direction.S turn.turn_right: 84988.0

direction.N turn.turn_left: 89196.0

direction.N turn.turn_right: 94420.0

direction.W turn.turn_left: 84796.0

direction.W turn.turn_right: 89060.0

direction.E turn.turn_left: 94452.0

direction.E turn.turn_right: 91460.0

Best result: 
direction.W turn.turn_left
84796.0
[[16 15 14 13 12]
 [17  4  3  2 11]
 [18  5  0  1 10]
 [19  6  7  8  9]
 [20 21 22 23 24]]


Mit der relativen Verschiebung des Startpunktes um ~(-0.019, -0.107) erhalten wir eine Verschlechterung des Scores zwischen den besten Spiralsuchen auf 84796 von 67376 und somit eine prozentuale Verschlechterung von ~26%

In [17]:
_ = analyse_spiral(spirals_r10, hist)

direction.S turn.turn_left: 74534.0

direction.S turn.turn_right: 70600.0

direction.N turn.turn_left: 72874.0

direction.N turn.turn_right: 77052.0

direction.W turn.turn_left: 70308.0

direction.W turn.turn_right: 72818.0

direction.E turn.turn_left: 76712.0

direction.E turn.turn_right: 75222.0

Best result: 
direction.W turn.turn_left
70308.0
[[16 15 14 13 12]
 [17  4  3  2 11]
 [18  5  0  1 10]
 [19  6  7  8  9]
 [20 21 22 23 24]]


Mit der relativen Verschiebung des Startpunktes um ~(0.026, -0.057) erhalten wir eine Verschlechterung des Scores zwischen den besten Spiralsuchen auf 70308 von 67376 und somit eine prozentuale Verschlechterung von ~4%

In [18]:
_ = analyse_spiral(spirals_r15, hist)

direction.S turn.turn_left: 65450.0

direction.S turn.turn_right: 62728.0

direction.N turn.turn_left: 63078.0

direction.N turn.turn_right: 65852.0

direction.W turn.turn_left: 62564.0

direction.W turn.turn_right: 63102.0

direction.E turn.turn_left: 65512.0

direction.E turn.turn_right: 65546.0

Best result: 
direction.W turn.turn_left
62564.0
[[16 15 14 13 12]
 [17  4  3  2 11]
 [18  5  0  1 10]
 [19  6  7  8  9]
 [20 21 22 23 24]]


Mit der relativen Verschiebung des Startpunktes um ~(0.015, -0.065) erhalten wir eine Verbesserung des Scores zwischen den besten Spiralsuchen auf 62564 von 67376 und somit eine prozentuale Verbesserung von ~7%

In [19]:
_ = analyse_spiral(spirals_r20, hist)

direction.S turn.turn_left: 60008.0

direction.S turn.turn_right: 61416.0

direction.N turn.turn_left: 62216.0

direction.N turn.turn_right: 61040.0

direction.W turn.turn_left: 62496.0

direction.W turn.turn_right: 64256.0

direction.E turn.turn_left: 61560.0

direction.E turn.turn_right: 61000.0

Best result: 
direction.S turn.turn_left
60008.0
[[20 19 18 17 16]
 [21  6  5  4 15]
 [22  7  0  3 14]
 [23  8  1  2 13]
 [24  9 10 11 12]]


Mit der relativen Verschiebung des Startpunktes um ~(0.019, -0.029) erhalten wir eine Verbesserung des Scores zwischen den besten Spiralsuchen auf 60008 von 67376 und somit eine prozentuale Verbesserung von ~11%.  
Somit liefert das 20er Raster die besten Ergebnisse.